### PR TITLE
added pr

### PR DESCRIPTION
**Pull request by Vagif12**
### Summary by CodeRabbit

**New Features:**
- Added a new GitHub Actions workflow, "Code Review", that triggers on pull requests and review comments.
- Introduced text normalization and cleaning functions in `c_test.py`.
- Implemented document type detection and content extraction in `main_model.py`.
- Enhanced `sliding_window_segmentation` function to escape characters before processing.

**Chore:**
- Removed the previous GitHub Actions workflow using the `code-review-gpt` package.

> 🎉 Here's to the code that's ever so bright, 🌟
> With workflows anew, it takes flight. 🚀
> Text is cleaned, accents stripped away, 📝
> Document types detected, without delay. 🕵️‍♀️
> A celebration of changes, oh what a sight! 🎆
> To the future of coding, we ignite. 🔥

### CODE REVIEW

**Comment by Vagif12** [comment]
Image description   CodeRabbit

## Walkthrough
This pull request introduces a new GitHub Actions workflow for automated code reviews using the `coderabbitai/ai-pr-reviewer` action, replacing the previous `code-review-gpt` package. It also adds new text processing and document handling functions in `c_test.py` and `main_model.py`, enhancing the application's capabilities with external libraries and APIs.

## Changes

| Files | Summary |
| --- | --- |
| `.github/workflows/ai-pr-reviewer.yml`, `.github/workflows/code-review-gpt.yml` | The PR adds a new GitHub Actions workflow "Code Review" that triggers on pull requests and review comments, utilizing the `coderabbitai/ai-pr-reviewer` action. Concurrently, it removes the old workflow that used the `code-review-gpt` package. |
| `c_test.py`, `main_model.py` | New functions are introduced for text normalization, character escaping, document type detection, and content extraction. These changes leverage external libraries and APIs for document processing, embeddings, vector stores, and question answering, without altering existing logic or functionality. |








---

<details>
<summary>Uplevel your code reviews with CodeRabbit Pro</summary>

### CodeRabbit Pro

If you like this project, please support us by purchasing the [Pro version](https://coderabbit.ai). The Pro version has advanced context, superior noise reduction and several proprietary improvements compared to the open source version.

</details>

**Review by Vagif12 — COMMENTED** [review]
Image description   CodeRabbit

<details>
<summary>Commits</summary>
Files that changed from the base of the PR and between 85da2dc16c7ac401b7b2eabc6451a0e033644395 and 9025eb8992aab795400e2f36a26bbd3329c4590e commits.
</details>

<details>
<summary>Files selected (3)</summary>

* .github/workflows/ai-pr-reviewer.yml (1)
* .github/workflows/code-review-gpt.yml (1)
* c_test.py (1)
</details>







<details>
<summary>Files skipped from review due to trivial changes (3)</summary>

* .github/workflows/ai-pr-reviewer.yml
* .github/workflows/code-review-gpt.yml
* c_test.py

</details>

<details>
<summary>Review comments generated (0)</summary>

* Review: 0
* LGTM: 0

</details>

---

<details>
<summary>Tips</summary>

### Chat with Image description  CodeRabbit Bot (`@coderabbitai`)
- Reply on review comments left by this bot to ask follow-up questions. A review comment is a comment on a diff or a file.
- Invite the bot into a review comment chain by tagging `@coderabbitai` in a reply.

### Code suggestions
- The bot may make code suggestions, but please review them carefully before committing since the line number ranges may be misaligned. 
- You can edit the comment made by the bot and manually tweak the suggestion if it is slightly off.

### Pausing incremental reviews
- Add `@coderabbitai: ignore` anywhere in the PR description to pause further reviews from the bot.

</details>

**Review by Vagif12 — COMMENTED** [review]
Image description   CodeRabbit

<details>
<summary>Commits</summary>
Files that changed from the base of the PR and between 9025eb8992aab795400e2f36a26bbd3329c4590e and cb2c2eb17c291ec7ca90e9a792f81dd60557f80b commits.
</details>

<details>
<summary>Files selected (1)</summary>

* main.py (1)
</details>





<details>
<summary>Files not summarized due to errors (1)</summary>

* main.py (diff tokens exceeds limit)

</details>


<details>
<summary>Files not reviewed due to errors (1)</summary>

* main.py (no response)

</details>


<details>
<summary>Review comments generated (0)</summary>

* Review: 0
* LGTM: 0

</details>

---

<details>
<summary>Tips</summary>

### Chat with Image description  CodeRabbit Bot (`@coderabbitai`)
- Reply on review comments left by this bot to ask follow-up questions. A review comment is a comment on a diff or a file.
- Invite the bot into a review comment chain by tagging `@coderabbitai` in a reply.

### Code suggestions
- The bot may make code suggestions, but please review them carefully before committing since the line number ranges may be misaligned. 
- You can edit the comment made by the bot and manually tweak the suggestion if it is slightly off.

### Pausing incremental reviews
- Add `@coderabbitai: ignore` anywhere in the PR description to pause further reviews from the bot.

</details>

File: main_model.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    image_types = ['jpg', 'jpeg', 'png', 'gif']
+    
+    if(guess_file.extension.lower() == "pdf"):
+        file_type = "pdf"
+        
+    elif(guess_file.extension.lower() in image_types):
+        file_type = "image"
+        
+    else:
+        file_type = "unkown"
```
Comment (Vagif12) [inline main_model.py:14]: Image description   CodeRabbit

There's a typo in the string "unknown". It should be corrected to ensure the function `detect_document_type` returns the correct value when the file type is not recognized.
```diff
-        file_type = "unkown"
+        file_type = "unknown"
```

File: main_model.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        
+    return file_type
+
+from langchain.document_loaders.image import UnstructuredImageLoader
+from langchain.document_loaders import UnstructuredFileLoader
```
Comment (Vagif12) [inline main_model.py:19]: Image description   CodeRabbit

The imports for `UnstructuredImageLoader` and `UnstructuredFileLoader` should be placed at the top of the file, following Python's best practices for import statements. This improves readability and maintainability.
```diff
- from langchain.document_loaders.image import UnstructuredImageLoader
- from langchain.document_loaders import UnstructuredFileLoader
```
Add these lines at the top of your script.

File: main_model.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+research_paper_chunks = text_splitter.split_text(research_paper_content)
+article_information_chunks = text_splitter.split_text(article_information_content)
+
+print(f"# Chunks in Research Paper: {len(research_paper_chunks)}")
+print(f"# Chunks in Article Document: {len(article_information_chunks)}")
+
+from langchain.embeddings.openai import OpenAIEmbeddings
+import os
+
+os.environ["OPENAI_API_KEY"] = "<YOUR_KEY>"
```
Comment (Vagif12) [inline main_model.py:68]: Image description   CodeRabbit

Hardcoding sensitive information like API keys directly into the code is a security risk. Consider using environment variables or secure configuration files to store this information.
```diff
- os.environ["OPENAI_API_KEY"] = "<YOUR_KEY>"
```
You can set the environment variable outside the script or use a secure method to load it.

File: main_model.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+from langchain.llms import OpenAI
+from langchain.chains.question_answering import load_qa_chain
+chain = load_qa_chain(OpenAI(), chain_type = "map_rerank",  
+                      return_intermediate_steps=True)
+
+def chat_with_file(file_path, query):
+    
+    file_content = extract_file_content(file_path)
+    text_splitter = text_splitter.split_text(file_content)
```
Comment (Vagif12) [inline main_model.py:89]: Image description   CodeRabbit

The variable `text_splitter` is being overwritten here. This could lead to unexpected behavior if you intend to use the original `text_splitter` later in the code. Consider renaming one of the variables to avoid confusion.
```diff
-    text_splitter = text_splitter.split_text(file_content)
+    text_chunks = text_splitter.split_text(file_content)
```
Then replace all subsequent uses of `text_splitter` with `text_chunks` in this function.

File: main_model.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    documents = document_search.similarity_search(query)
+    
+    results = chain({
+                        "input_documents":documents, 
+                        "question": query
+                    }, 
+                    return_only_outputs=True)
+    answers = results['intermediate_steps'][0]
+    
+    return answers
```
Comment (Vagif12) [inline main_model.py:101]: Image description   CodeRabbit

It's good practice to end your files with a newline character to comply with POSIX standards and ensure compatibility across different platforms and tools.
```diff
+ \n
```
Just add a newline at the end of the file.

**Review by Vagif12 — COMMENTED** [review]
Image description   CodeRabbit

<details>
<summary>Commits</summary>
Files that changed from the base of the PR and between cb2c2eb17c291ec7ca90e9a792f81dd60557f80b and f8ce666af8cdaa7c51a8f2873b8dd836cf17edff commits.
</details>

<details>
<summary>Files selected (1)</summary>

* main_model.py (1)
</details>








<details>
<summary>Review comments generated (5)</summary>

* Review: 5
* LGTM: 0

</details>

---

<details>
<summary>Tips</summary>

### Chat with Image description  CodeRabbit Bot (`@coderabbitai`)
- Reply on review comments left by this bot to ask follow-up questions. A review comment is a comment on a diff or a file.
- Invite the bot into a review comment chain by tagging `@coderabbitai` in a reply.

### Code suggestions
- The bot may make code suggestions, but please review them carefully before committing since the line number ranges may be misaligned. 
- You can edit the comment made by the bot and manually tweak the suggestion if it is slightly off.

### Pausing incremental reviews
- Add `@coderabbitai: ignore` anywhere in the PR description to pause further reviews from the bot.

</details>

**Comment by Vagif12** [comment]
@coderabbitai can you write test cases for the chat_with_file method?

File: main_model.py (hunk 5 above)
Comment (Vagif12) [reply main_model.py:101]: @coderabbitai can you write test cases for the chat_with_file method?

Comment (Vagif12) [reply main_model.py:101]: Image description   CodeRabbit

@Vagif12 Sure, here are some basic test cases for the `chat_with_file` method. These tests use Python's built-in `unittest` module.

```python
import unittest
from unittest.mock import patch, MagicMock

class TestChatWithFile(unittest.TestCase):
    @patch('main_model.extract_file_content')
    @patch('main_model.get_doc_search')
    @patch('main_model.load_qa_chain')
    def test_chat_with_file(self, mock_load_qa_chain, mock_get_doc_search, mock_extract_file_content):
        # Mocking the dependencies
        mock_extract_file_content.return_value = "This is a sample file content"
        mock_get_doc_search.return_value = MagicMock()
        mock_load_qa_chain.return_value = MagicMock()

        # Define a sample query
        query = "What is the content of the file?"

        # Call the function with the sample file path and query
        result = chat_with_file("sample_file_path", query)

        # Check if the dependencies were called correctly
        mock_extract_file_content.assert_called_once_with("sample_file_path")
        mock_get_doc_search.assert_called_once()
        mock_load_qa_chain.assert_called_once()

        # Add more assertions based on your specific requirements and expected results

if __name__ == '__main__':
    unittest.main()
```

Please note that this is a basic test case and it assumes that the `chat_with_file` function is in the global scope. If the function is part of a class, you might need to adjust the test case accordingly. Also, you may want to add more assertions based on your specific requirements and expected results.

File: main_model.py (hunk 4 above)
Comment (Vagif12) [reply main_model.py:89]: @coderabbitai can you add test cases?

Comment (Vagif12) [reply main_model.py:89]: Image description   CodeRabbit

@Vagif12 Sure, here are some basic test cases for the functions in `main_model.py`. 

```python
import unittest

class TestMainModel(unittest.TestCase):

    def test_detect_document_type(self):
        self.assertEqual(detect_document_type("./data/test.pdf"), "pdf")
        self.assertEqual(detect_document_type("./data/test.jpg"), "image")
        self.assertEqual(detect_document_type("./data/test.txt"), "unknown")

    def test_extract_file_content(self):
        # Assuming that the files contain the following content
        self.assertEqual(extract_file_content("./data/test.pdf"), "This is a test PDF document.")
        self.assertEqual(extract_file_content("./data/test.jpg"), "This is a test image document.")

    def test_chat_with_file(self):
        # Assuming that the file contains the question "What is AI?" and the answer "AI stands for Artificial Intelligence."
        self.assertEqual(chat_with_file("./data/test.pdf", "What is AI?"), "AI stands for Artificial Intelligence.")

if __name__ == '__main__':
    unittest.main()
```

Please note that these tests are quite simple and may not cover all edge cases. They also assume that certain files exist with specific content. You might want to create those files or replace the paths with the paths of your actual test files.

Also, please remember to replace the hardcoded answers in the `test_chat_with_file` function with the expected output based on the content of your test files.
